File: pass-by-value-pass-by-reference.c
#include<stdio.h>

typedef struct {
    int num;
} OBJ_TYPE;

void receiveByValue(int v) { // this will make a copy, it will create a copy in another memory address
    v += 5;
}

void receiveByReference(int *r) { //this not copy, it will point to original memory address
    *r += 5;
}

void receiveStructByValue(OBJ_TYPE obj) {
    obj.num = 5;
}

int main() {
    int v = 0;
    receiveByValue(v);
    printf("It will print 'zero': %i \n", v);

    int r = 0;
    receiveByReference(&r);
    printf("It will print 'five': %i \n", r);

    OBJ_TYPE obj;
    obj.num = 0;

    receiveStructByValue(obj);
    printf("It will print 'zero': %i \n", obj.num);

    return 0;
}

File: pass-by-value-pass-by-reference.js
function receiveObjectsType(arr, obj) { //this not copy, it will point to original memory address
    arr.push('foo');
    obj.name = 'alencar';
}

function receivePrimitiveType(primiveType) { // this will make a copy, it will create a copy in another memory address
    primiveType = primiveType + 5;
}

function main() {
    var arr = [];
    var primiveType = 0;
    var obj = {
        name: 'thais'
    }

    receiveObjectsType(arr, obj);
    receivePrimitiveType(primiveType);
    
    console.log(`It will print 'foo': ${arr}`);
    console.log(`It will print '0': ${primiveType}`);
    console.log(`It will print 'alencar': ${obj.name}`);
}

main();